# Kalkulation der Samplesize

## Model 1

Wir laden in einem ersten Schritt die Daten.


In [9]:
import os
from pathlib import Path
import pandas as pd

target = Path("C:/Users/mhu/Documents/github/topic_model_it")
os.chdir(target)
print(f"Aktuelles Arbeitsverzeichnis: {os.getcwd()}")

df = pd.read_csv("data/informatikkurse.csv")
print(df.head())

Aktuelles Arbeitsverzeichnis: C:\Users\mhu\Documents\github\topic_model_it
   Unnamed: 0  veranstaltung_id  \
0           1           1717848   
1           2           1717914   
2           3           1717915   
3           4           1717989   
4           5           1717990   

                                 veranstaltung_titel kursbeschreibung  \
0    Übung zu: Didaktische Grundlagen für Informatik              NaN   
1  Seminar zu: Lehren und Lernen im Fach Informat...              NaN   
2                  Übung zu: Ethik in der Informatik              NaN   
3                  Übung zu: Ethik in der Informatik              NaN   
4                  infSP2F-01a: Softwareprojekt (2F)              NaN   

  kursformat_original sprache_original lernziele lernmethode literatur  \
0               Übung              NaN       NaN         NaN       NaN   
1             Seminar              NaN       NaN         NaN       NaN   
2               Übung              NaN       NaN     

C:\Users\mhu\AppData\Local\Temp\ipykernel_13852\4191572470.py:9: DtypeWarning: Columns (15,16) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("data/informatikkurse.csv")


### Sample-Size

In einem weiteren Schritt kalkulieren wir die benötigte Größe der gelabelten Daten.

In [10]:
import math

def calculate_sample_size(N=None, p=0.5, z=1.96, e=0.03):
    """
    Berechnet die benötigte Stichprobengröße basierend auf der Cochran-Formel.
    
    Parameter:
    N: Gesamtpopulation (Optional). Falls None, wird n_0 berechnet.
    p: Erwarteter Anteil (Default 0.5 für maximale Stichprobe).
    z: Z-Wert (Default 1.96 für 95% Konfidenz).
    e: Fehlermarge (Default 0.03 für 3%).
    """
    
    # 1. Basis-Stichprobengröße (n0) berechnen
    n_0 = (z**2 * p * (1 - p)) / (e**2)
    
    if N is None:
        return math.ceil(n_0)
    
    # 2. Korrekturterm für kleine Populationen berechnen
    # Korrekturterm = 1 + (z^2 * p * (1-p)) / (e^2 * N)
    correction_term = 1 + (n_0 / N)
    
    # Finale Stichprobengröße = n0 / Korrekturterm
    sample_size = n_0 / correction_term
    
    return math.ceil(sample_size)

# Beispiel aus deinem Dokument: N = 4.000.000, e = 3%, p = 50%, z = 1.96
len_data = len(df)
result = calculate_sample_size(N=len_data, p=0.9, z=1.96, e=0.03)
print(f"Benötigte Samples zum Labeln: {result}")

Benötigte Samples zum Labeln: 382


Die Benötigte Größe der gelabelter Daten beträgt für Model 1: `n=382`